### PR TITLE
8289697: buffer overflow in MTLVertexCache.m: MTLVertexCache_AddGlyphQuad

### PR DESCRIPTION
**Pull request by VladimirKempik**
Please review this simple patch. When running idea on jdk17 with asan I have found this buffer overflow.
The code checks the cache for at least one free element, while placing 6 elements to the cache.
The fix checks the presence of 6 free elements.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289697](https://bugs.openjdk.org/browse/JDK-8289697): buffer overflow in MTLVertexCache.m: MTLVertexCache_AddGlyphQuad

### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to [2676028f](https://git.openjdk.org/jdk/pull/9368/files/2676028f80505e53a5a4463c388b2465a4a5eeee)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9368/head:pull/9368` \
`$ git checkout pull/9368`

Update a local copy of the PR: \
`$ git checkout pull/9368` \
`$ git pull https://git.openjdk.org/jdk pull/9368/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9368`

View PR using the GUI difftool: \
`$ git pr show -t 9368`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9368.diff">https://git.openjdk.org/jdk/pull/9368.diff</a>

</details>

### CODE REVIEW

**Comment by avu** [comment]
Looks good for me

**Comment by VladimirKempik** [comment]
Linux test failure unrelated, this commit changes only macos's code

**Review by aghaisas — COMMENTED** [review]
This is a good catch and fix.

There is another place in this file where `MTLVC_ADD_TRIANGLES` is used.
Although it is a corner case, it is worth adding the check for additional 6 vertices at that place as well.
Adding a separate check at the beginning of the `MTLVertexCache_AddMaskQuad` method where there is a check for `maskCacheIndex` seems logical to me. What do you think?

**Comment by VladimirKempik** [comment]
Added one more check into MTLVertexCache_AddMaskQuad,  to the first if condition.

**Review by prrace — COMMENTED** [review]
Are there any other cases like this ? We should look around. (PS I see Ajit asked the same question and perhaps even answered it !)
Did you find this by inspection or did you hit it ?
Can we use a defined constant instead of "6" ?

**Comment by VladimirKempik** [comment]
> Are there any other cases like this ? We should look around. (PS I see Ajit asked the same question and perhaps even answered it !) Did you find this by inspection or did you hit it ? 

I hit this bug when running IDEA on asan-enabled build of ojdk
>Can we use a defined constant instead of "6" ?

#define TRI_IN_VERT 6 ?

File: src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLVertexCache.m [hunk 1]
```diff
@@ -65,6 +65,9 @@
         MTLVC_ADD_VERTEX(TX1, TY1, DX1, DY1, 0); \
     } while (0)
 
+//Next define should exactly match to the amount of MTLVC_ADD_VERTEX in MTLVC_ADD_TRIANGLES
+#define MTL_TRIS_IN_VERTEX 6
```
Comment (aghaisas) [inline src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLVertexCache.m:69]: Constants starting with `MTL_` should be avoided as they are used in Metal Framework library.
I think, this should be `VERTS_FOR_A_QUAD`.

**Comment by aghaisas** [comment]
> Yeah, missed that, thanks for noticing

There is one instance of `MTL_TRIS_IN_VERTEX` remaining in a comment!

**Review by prrace — APPROVED** [review]
LGTM although I'd really (really!) like you to break up those long lines before pushing.

**Pull request closed without merging**